### ISSUE TITLE
ValidateSamFile should return more informative error codes

### ISSUE THREAD

**Issue by vdauwera** [broadinstitute/picard#794]
### Tool(s) involved
ValidateSamFile

### Description
ValidateSamFile returns 1 when a file fails validation and when the tool crashes. The tool should return different codes so that we can differentiate "tool crashed" from "file has some problems". Ideally we'd want to further distinguish "there were some WARNs" from "there were some ERRORs".

This is important for running the tool in pipelining environments like FireCloud.

**Comment by ronlevine** [comment]
ValidateSamFile calls the HTSJDK methods SamFileValidator#validateSamFileSummary and SamFileValidator#validateSamFileVerbose, which return boolean values to indicate validation errors. The validation errors are stored as an enum, SAMValidationError.Type. These methods need to return the enum's ordinal. 
@vdauwera If a numerical value is returned, how would the pipeline know what it corresponds to? Should an error message be added to each SAMValidationError.Type value?

**Comment by tfenne** [comment]
@ronlevine Maybe I'm missing something, but I think @vdauwera was asking for something simpler - that if the file does not validate that the tool exit with any value other than 0 or 1, so you might end up with:
  `exit(0)` == successfully read whole file and it was all valid
  `exit(1)` == the program failed to complete execution (exception, IO issue, etc.)
  `exit(2)` == successfully read the whole file but there were validation warnings
  `exit(3)` == successfully read the whole file but there were validation errors

With separate exits for validation warnings/errors being a nice to have.

**Comment by ronlevine** [comment]
@tfenne I might be missing something. I am not familiar with the use cases for this tool within the pipelines. If those exit cases are as fine grained as needed, I'll go with that.

**Comment by vdauwera** [comment]
Yep, @tfenne has it right, that's all we would need. Sorry I wasn't clearer in the original description. 

In the case of Cromwell we can set the execution engine to evaluate the return code intelligently, so it can correctly identify that the tool execution succeeded, yet still convey programmatically that the data file was not fully valid.

**Comment by ronlevine** [comment]
HTSJDK's `SamFileValidator` needs to keep track of warnings and errors. `ValidateSamFile` can return the following error codes:
-1  - `program failed to complete execution`
0   - `program ran successfully`
1   - `validation warnings but no errors`
2  - `validation errors and possibly warnings`

**Comment by ronlevine** [comment]
Changes in SamFileValidator are blocked until htsjdk > 2.9.1 is released.

**Comment by ronlevine** [comment]
Unblocking by using the latest htsjdk snapshot.

**Comment by ronlevine** [comment]
Closing since #822 is merged.
